### PR TITLE
Add maximum marking

### PR DESCRIPTION
**Pull request by schnellerhase**
Implements a maximum marking criterion for AFEM schemes, i.e. for $\theta \in (0, 1)$ and a marker $\eta \in \mathbb{R}^N$, computes

$$
  \mathcal{I} = \left\\{ i \ | \ \eta_i > \theta \max_j \eta_j \right\\}.
$$

In particular this introduces/defines the interface for other marking routines (to follow in the future): based on a process local list of (entity-)indicators the (local) marked index set of it is computed.

<details>

<summary>
Test program to visualise results (parallel ready).
</summary>

```python
from mpi4py import MPI

import numpy as np

from dolfinx import mesh, plot
import pyvista as pv

mesh_0 = mesh.create_unit_square(comm := MPI.COMM_WORLD, n := 10, n)

cell_im = mesh_0.topology.index_map(tdim := mesh_0.topology.dim)
cell_count = cell_im.size_local + cell_im.num_ghosts

marker = np.random.random(cell_count)
marked_cells = mesh.mark_maximum(marker, theta := 0.5, comm)

mesh_0.topology.create_entities(1)
marked_edges = mesh.compute_incident_entities(mesh_0.topology, marked_cells, tdim, 1)

mesh_1, _, _ = mesh.refine(mesh_0, marked_edges)

grid_0 = pv.UnstructuredGrid(*plot.vtk_mesh(mesh_0))
grid_1 = pv.UnstructuredGrid(*plot.vtk_mesh(mesh_1))

plotter = pv.Plotter(shape=(1, 2))
plotter.subplot(0, 0)
plotter.add_mesh(grid_0, show_edges=True)
plotter.view_xy()
plotter.add_text("mesh_0", position="upper_left")

plotter.subplot(0, 1)
plotter.add_mesh(grid_1, show_edges=True)
plotter.view_xy()
plotter.add_text("mesh_1", position="upper_left")

plotter.link_views()
plotter.show()

```
</details>

Work towards https://github.com/FEniCS/dolfinx/issues/4141.

### CODE REVIEW

File: cpp/dolfinx/refinement/mark.h [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+/// entity
+/// @param[in] theta Cut off parameter, 0 ≤ θ ≤ 1
+/// @param[in] comm Communicator over which the maximum is computed.
+/// @return Indices (local) of marker elements, which satisfy: marker_i ≥ θ
+/// max(marker).
+template <std::floating_point T>
+std::vector<std::int32_t> mark_maximum(std::span<const T> marker, T theta,
+                                       MPI_Comm comm)
+{
+  assert((0 <= theta) && (theta <= 1));
```
Comment (jhale) [inline cpp/dolfinx/refinement/mark.h]: Cheap and useful check - should always throw std::value_error

File: cpp/dolfinx/refinement/mark.h [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  T max = marker.empty() ? std::numeric_limits<T>::min()
+                         : std::ranges::max(marker);
+  MPI_Allreduce(MPI_IN_PLACE, &max, 1, dolfinx::MPI::mpi_t<T>, MPI_MAX, comm);
+
+  auto mark = [=](auto e) { return e >= theta * max; };
+
+  std::vector<std::int32_t> indices;
+  indices.reserve(std::ranges::count_if(marker, mark));
+
+  for (std::int32_t i = 0; i < static_cast<std::int32_t>(marker.size()); ++i)
```
Comment (jhale) [inline cpp/dolfinx/refinement/mark.h:48]: Is there a std library way to do this loop?

File: cpp/dolfinx/refinement/mark.h [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+std::vector<std::int32_t> mark_maximum(std::span<const T> marker, T theta,
+                                       MPI_Comm comm)
+{
+  assert((0 <= theta) && (theta <= 1));
+
+  T max = marker.empty() ? std::numeric_limits<T>::min()
+                         : std::ranges::max(marker);
+  MPI_Allreduce(MPI_IN_PLACE, &max, 1, dolfinx::MPI::mpi_t<T>, MPI_MAX, comm);
+
+  auto mark = [=](auto e) { return e >= theta * max; };
```
Comment (jhale) [inline cpp/dolfinx/refinement/mark.h]: Use const T instead of auto e?

Also remove inequality, pointless for floating point arithmetic.

**Review by jhale — DISMISSED** [review]

File: cpp/dolfinx/refinement/mark.h (hunk 2 above)
Comment (schnellerhase) [reply cpp/dolfinx/refinement/mark.h:48]: Yes, but only in C++23 https://en.cppreference.com/cpp/ranges/enumerate_view.

File: cpp/dolfinx/refinement/mark.h (hunk 3 above)
Comment (schnellerhase) [reply cpp/dolfinx/refinement/mark.h]: Changed to `T` - think we do not follow const correctness for trivial types.

Comment (schnellerhase) [reply cpp/dolfinx/refinement/mark.h]: > Also remove inequality, pointless for floating point arithmetic.

Don't think this is correct: for $\theta=1.0$ and $\eta = [ 1, \dots, 1]$ that would then mark nothing (tested).

File: cpp/dolfinx/refinement/mark.h [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+/// @return Indices (local) of marker elements, which satisfy: marker_i ≥ θ
+/// max(marker).
+template <std::floating_point T>
+std::vector<std::int32_t> mark_maximum(std::span<const T> marker, T theta,
+                                       MPI_Comm comm)
+{
+  if ((theta < 0) || (theta > 1))
+    throw std::invalid_argument("Theta needs to fullfill 0 ≤ θ ≤ 1.");
+
+  T max = marker.empty() ? std::numeric_limits<T>::min()
```
Comment (chrisrichardson) [inline cpp/dolfinx/refinement/mark.h]: Does this make sense? Shouldn't it just be zero?

File: cpp/dolfinx/refinement/mark.h [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+                         : std::ranges::max(marker);
+  MPI_Allreduce(MPI_IN_PLACE, &max, 1, dolfinx::MPI::mpi_t<T>, MPI_MAX, comm);
+
+  auto mark = [=](T e) { return e >= theta * max; };
```
Comment (chrisrichardson) [inline cpp/dolfinx/refinement/mark.h:43]:
```suggestion
  auto mark = [=](T e) { return e > theta * max; };
```

File: cpp/dolfinx/refinement/mark.h (hunk 4 above)
Comment (schnellerhase) [reply cpp/dolfinx/refinement/mark.h]: We do not assume the marker to be positive, so smaller than zero values are admissible in the marker, therefore 0 would not allow for the correct max value being identified in the all reduce.

File: cpp/dolfinx/refinement/mark.h (hunk 5 above)
Comment (schnellerhase) [reply cpp/dolfinx/refinement/mark.h:43]: Ref https://github.com/FEniCS/dolfinx/pull/4156#discussion_r3109060529.

File: cpp/dolfinx/refinement/mark.h (hunk 4 above)
Comment (chrisrichardson) [reply cpp/dolfinx/refinement/mark.h]: It depends what you are using for `T` - but for most floating point types, this will just be a small positive number:
https://en.cppreference.com/cpp/types/numeric_limits/min

Comment (schnellerhase) [reply cpp/dolfinx/refinement/mark.h]: Right, should be `lowest` - thanks!

**Review by chrisrichardson — APPROVED** [review]

**Comment by jhale** [comment]
My comments on the floating point comparisons are not resolved.

File: cpp/dolfinx/refinement/mark.h (hunk 3 above)
Comment (jhale) [reply cpp/dolfinx/refinement/mark.h]: And through what astronomically unlikely event would eta be binary 1.0 for every cell?

**Comment by chrisrichardson** [comment]
@jhale - I think @schnellerhase did address it. Did you mean that it should be a simple `>` rather than `>=`?

**Comment by jhale** [comment]
Yes - the equality is completely redundant in a floating point context and the example given is a contrived one.

File: cpp/dolfinx/refinement/mark.h (hunk 3 above)
Comment (schnellerhase) [reply cpp/dolfinx/refinement/mark.h]: Unlikely but valid input to the function. We can either leave as is or change to $e + \epsilon > \theta \max \eta$ for some close to machine epsilon $\epsilon$, which should ensure this edge case passes. Fine with both.

File: python/test/unit/refinement/test_mark.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    marked_cells = mesh.mark_maximum(marker, theta, comm)
+
+    assert np.allclose(
+        marked_cells,
+        np.argwhere(marker >= theta * comm.allreduce(np.max(marker), MPI.MAX)).flatten(),
+    )
+
+    msh.topology.create_entities(1)
+    marked_edges = mesh.compute_incident_entities(msh.topology, marked_cells, tdim, 1)
+    mesh.refine(msh, marked_edges)
```
Comment (jorgensd) [inline python/test/unit/refinement/test_mark.py:33]: Should there be something more at the end of this test ?

Comment (schnellerhase) [reply python/test/unit/refinement/test_mark.py:33]: I just wanted to check that the input works as expected with the refine calls - so the test being "refine works". Not sure what to exactly test for on the refined mesh, up to it being a mesh.

File: cpp/dolfinx/refinement/mark.h [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+{
+  if ((theta < 0) || (theta > 1))
+    throw std::invalid_argument("Theta needs to fullfill 0 ≤ θ ≤ 1.");
+
+  T max = marker.empty() ? std::numeric_limits<T>::lowest()
+                         : std::ranges::max(marker);
+  MPI_Allreduce(MPI_IN_PLACE, &max, 1, dolfinx::MPI::mpi_t<T>, MPI_MAX, comm);
+
+  auto mark = [=](T e)
+  { return e + std::numeric_limits<T>::epsilon() * 1e2 > theta * max; };
```
Comment (jhale) [inline cpp/dolfinx/refinement/mark.h]: You're overthinking this.

Why use epsilon around 1 when e and max can vary across many orders of magnitude? Why 1e-2? Why systematically favor inclusion over exclusion?

Why does this matter in the context of mesh refinement where small perturbations in the marked set make no difference?

**Review by jhale — CHANGES_REQUESTED** [review]

File: cpp/dolfinx/refinement/mark.h (hunk 7 above)
Comment (schnellerhase) [reply cpp/dolfinx/refinement/mark.h]: Since the maximum is attained by a value of the marker, we need inclusion. Otherwise there is no guarantee that an element gets marked for $\theta = 1$.

This is now a less than check with abs. tol of 1e2 * machine eps tolerance to make this check as tight as possible while allowing some reasonable leeway for range differences. Can adjust tolerances or change to relative check.

Comment (schnellerhase) [reply cpp/dolfinx/refinement/mark.h]: I think if we restrict to $\theta \in (0,1)$ we can stick to strict inequality checks.

Comment (jhale) [reply cpp/dolfinx/refinement/mark.h]: I agree - the endpoint at 1.0 is best done through the obvious approach (refine everything).

File: python/test/unit/refinement/test_mark.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+    tdim = msh.topology.dim
+    cell_count = (cell_im := msh.topology.index_map(tdim)).size_local + cell_im.num_ghosts
+    marker = np.random.default_rng(0).random(cell_count)
+
+    marked_cells = mesh.mark_maximum(marker, theta, comm)
+
+    assert np.allclose(
+        marked_cells,
+        np.argwhere(marker >= theta * comm.allreduce(np.max(marker), MPI.MAX)).flatten(),
```
Comment (schnellerhase) [inline python/test/unit/refinement/test_mark.py]:
```suggestion
        np.argwhere(marker > theta * comm.allreduce(np.max(marker), MPI.MAX)).flatten(),
```

**Review by jhale — APPROVED** [review]
